Fixes broadcast spill serialization/deserialization [databricks]

abellina opened the pull request:
Closes https://github.com/NVIDIA/spark-rapids/issues/8602
Closes https://github.com/NVIDIA/spark-rapids/issues/8603

This PR fixes a couple of issues in broadcast related to Spark's `MemoryStore` spill and due to us marking the `SerializeConcatHostBuffersDeserializeBatch` as `AutoCloseable` (this is no longer the case as of this PR).

We change the input of `SerializeConcatHostBuffersDeserializeBatch` to be:  `HostConcatResult`, `numRows`, and `dataLen`. `HostConcatResult` is null in the case we have a rows-only broadcast. Once the batch is materialized on the GPU, we defer to the materialized spillable as the source of truth for the various methods, including `hostBatch`. I've added tests to cover the various spill scenarios (before GPU materialization and after GPU materialization) + obtaining a `hostBatch` from driver or the executor.

abellina commented:
build

abellina commented:
build

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala:
```diff
@@ -151,93 +162,146 @@ class SerializeConcatHostBuffersDeserializeBatch(
+ */
 @SerialVersionUID(100L)
 class SerializeBatchDeserializeHostBuffer(batch: ColumnarBatch)
   extends Serializable with AutoCloseable {
```
This is still AutoCloseable -- intentional?

abellina replied:
This is AutoCloseable but it doesn't end up mattering since it's not the broadcast result. I can make it not AutoCloseable and instead implement closeInternal to make it clear...

abellina replied:
So I do end up using `.close` in this function https://github.com/NVIDIA/spark-rapids/pull/8617/files#diff-7ca82be5615724f3ba4376b82be1c8a34e3fb5df34a3dedfa9a21eb3e6e54b97R617, when creating a `SerializeConcatHostBuffersDeserializeBatch` from an array of `SerializeBatchDeserializeHostBuffer`

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala:
```diff
@@ -90,56 +88,69 @@ object SerializedHostTableUtils {
   // used for memoization of deserialization to GPU on Executor
   @transient private var batchInternal: SpillableColumnarBatch = null
 
+  private def maybeGpuBatch: Option[SpillableColumnarBatch] = Option(batchInternal)
+
```
Because of `Serializable`.

On deserialization, a `batchInternal: Option[SpillableColumnarBatch]` will not be `None`, it will be `Null`. So, I thought this might be a bit cleaner.

abellina commented:
I see dpp_test failed (I was running other suites locally) and it is related. I will debug and patch up.

abellina commented:
> I see dpp_test failed (I was running other suites locally) and it is related. I will debug and patch up.

Should be fixed by https://github.com/NVIDIA/spark-rapids/pull/8617/commits/48e8c054f55a77bfb0618f661595818d0aed9143. 

I was closing a `hostBatch` result too soon since we use a row iterator to project out in the `GpuSubqueryBroadcastExec`. I moved the .toArray within the `withResource` and added a comment.

abellina commented:
build

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala:
```diff
@@ -151,93 +162,146 @@ class SerializeConcatHostBuffersDeserializeBatch(
 }
 // scalastyle:on no.finalize
 
+// scalastyle:off no.finalize
```
nit: Why turn it on to turn it off right afterwards???

abellina replied:
I can make that one scalastyle:off block, it was mostly so that it was clear that it was needed in both classes.

revans2 reviewed (approved):
The code changes look fine to me for the most part.

jlowe reviewed (approved):
lgtm.  We can address the cleanup for driver concat and simplification/symmetry in a followup.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala:
```diff
@@ -90,56 +88,69 @@ object SerializedHostTableUtils {
+ */
 // scalastyle:off no.finalize
 @SerialVersionUID(100L)
 class SerializeConcatHostBuffersDeserializeBatch(
```
It is not about the name, but that is part of it. It is the fact that we have multiple different Serializable classes that take one type of input and produce a different type of output. Why? The original reason was because we needed to broadcast an object from the host and put it on the device, where it would be deduped on the device. But is that really saving us anything. Would it be better to just send/receive a HostBuffer and the headers.  Just one Serializable class that does that. Then everything else are helper functions that transform it in different ways.

Not that this blocks the current patch from going in. I just thought that was the direction we were going to try, but I guess I was wrong.

abellina replied:
I started to file a follow on but I am not entirely sure how I'd merge the two serializables into one, well I can think of one way. Here's what happens today:

1. Every partition produces a `SerializeBatchDeserializeHostBuffer` from the executors. This class has to be serializable takes a `ColumnarBatch` and it is able to write and read from the streams.
2. The driver collects `Array[SerializeBatchDeserializeHostBuffer]`, where each element is a partition.
3. The driver concats this array and produces `HostConcatResult`.
4. The driver wraps the host concatenated result using `SerializeConcatHostBuffersDeserializeBatch` and broadcasts this out. 
5. Executors deal with the `HostConcatResult` only.

In order to remove the number of serializable classes, I think we could broadcast the Array that we collected in (2) as `Array` is `Serializable` and then the executors would concatenate these results when materializing the batch:

1. Every partition produces a `SerializeBatchDeserializeHostBuffer` from the executors. This class has to be serializable takes a `ColumnarBatch` and it is able to write and read from the streams.
2. The driver collects `Array[SerializeBatchDeserializeHostBuffer]`, where each element is a partition.
3. The driver broadcasts the `Array` directly.
4. Executors deal with serializing and deserializing this Array upto materialization. 
5. At materialiation, the executors create a `HostConcatResult` and turn it into a spillable GPU batch.

abellina replied:
ok I synced with @revans2 here and the main comment here was that the driver could still concatenate but then broadcast out essentially the `SerializeBatchDeserializeHostBuffer` again (as the result of the concatenation is a header and a host buffer, so this object could be reused). This would help with simplifying the code and make it easier for us to go to the next step to improve Broadcast which was to allow these broadcast GPU `RapidsBuffer` to be freed instead of actually spilled, since they are already in host memory at the source. We can work on this cleanup as part of https://github.com/NVIDIA/spark-rapids/issues/7709.

abellina commented:
Thanks for the reviews @jlowe and @revans2!
